common/auto-updater/versions-db-manager.cpp

```diff
@@ -0,0 +1,307 @@
+                            }
+                        }
+                        else
+                        {
+                            LOG_WARNING("Server versions file parsing error - " + element_key + " all values should be represented as a string \n");
+                            throw std::runtime_error("Server versions file parsing error - " + element_key + " - unknown field \n");
+                        }
+                    }
+                    else
+                    {
+                        LOG_WARNING("Server versions file parsing error - " + element_key + " - unknown field \n");
+                        throw std::runtime_error("Server versions file parsing error - " + element_key + "- unknown field \n");
+                    }
+                }
+
+                // Verify each key in the schema can be found in the current version fields
+                //std::pair<const K, E> &item
+                if (std::all_of(schema.cbegin(), schema.cend(), [curr_version](const std::pair<std::string, std::function<bool(const std::string&)>> &schema_item)
+                {
+                    return curr_version.end() != std::find_if(curr_version.cbegin(), curr_version.cend(), [schema_item](const std::pair<std::string, std::string> &ver_item) {return schema_item.first == ver_item.first; });
+                }))
+                {
+                    _server_versions_vec.emplace_back(curr_version); // Version added to valid versions vector
+                }
+                else
+                {
+                    LOG_WARNING("Server versions json file corrupted - not matching schema requirements \n");
+                    throw std::runtime_error("Server versions json file corrupted - not matching schema requirements \n");
+                }
+            }
+        }
+        else
+        {
+            LOG_WARNING("Server versions json file corrupted - Title not as expected\n");
+            throw std::runtime_error("Server versions json file corrupted - Title not as expected\n");
+        }
+    }
+
+    const std::string versions_db_manager::get_platform() const
+    {
+        std::string platform;
+
+#ifdef _WIN64
+        platform = "Windows amd64";
+
+#elif _WIN32
+        platform = "Windows x86";
+#elif __linux__ 
+#ifdef __arm__
+        platform = "Linux arm";
+#else
+        platform = "Linux amd64";
+#endif
+#elif __APPLE__
+        platform = "Mac OS";
+#elif __ANDROID__
+        platform = "Linux arm";
+#endif
+        return platform;
+    }
+
+    bool versions_db_manager::init()
+    {
+        // Load server versions info on first access
+        if (!_server_versions_loaded)
+        {
+            std::stringstream server_versions_data;
+            server_versions_data.clear();
+            // Download / Open the json file
+            if (!get_server_data(server_versions_data))
```

Comment:
Maybe you should add a counter , so if after 10-TBD failed iteration it will give up and stop additional waits as it will annoy the user

Reply:
This function Is throw able + has a return value.
IMHO - If the user wants to call it in a loop / several times he should implement this logic if he likes this behavior. the service class should not prevent it.